subsys/shell/shell_telnet.c

```diff
@@ -0,0 +1,493 @@
+{
+	/* Commands are two or three bytes. */
+	NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(cmd_access, u16_t);
+	struct telnet_simple_command *cmd;
+
+	cmd = (struct telnet_simple_command *)net_pkt_get_data(pkt,
+							       &cmd_access);
+	if (!cmd || cmd->iac != NVT_CMD_IAC) {
+		return false;
+	}
+
+#ifdef CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND
+	LOG_DBG("Got a command %u/%u/%u", cmd->iac, cmd->op, cmd->opt);
+	telnet_reply_command(cmd);
+#endif  /* CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND */
+
+	return true;
+}
+
+static void telnet_recv(struct net_context *client,
+			struct net_pkt *pkt,
+			union net_ip_header *ip_hdr,
+			union net_proto_header *proto_hdr,
+			int status,
+			void *user_data)
+{
+	size_t len;
+
+	if (!pkt || status) {
+		telnet_end_client_connection();
+
+		LOG_DBG("Telnet client dropped (AF_INET%s) status %d",
+			net_context_get_family(client) == AF_INET ?
+			"" : "6", status);
+		return;
+	}
+
+	len = net_pkt_remaining_data(pkt);
+	if (len < TELNET_MIN_MSG) {
+		LOG_DBG("Packet smaller than minimum length");
+		goto unref;
+	}
+
+	if (telnet_handle_command(pkt)) {
+		LOG_DBG("Handled command");
+		goto unref;
+	}
+
+	/* Fifo add */
+	k_fifo_put(&sh_telnet->rx_fifo, pkt);
+
+	sh_telnet->shell_handler(SHELL_TRANSPORT_EVT_RX_RDY,
+				 sh_telnet->shell_context);
+
+	return;
+
+unref:
+	net_pkt_unref(pkt);
+}
+
+static void telnet_accept(struct net_context *client,
+			  struct sockaddr *addr,
+			  socklen_t addrlen,
+			  int error,
+			  void *user_data)
+{
+	if (error) {
+		LOG_ERR("Error %d", error);
+		goto error;
+	}
+
+	if (sh_telnet->client_ctx) {
+		LOG_WRN("A telnet client is already in.");
+		goto error;
+	}
+
+	if (net_context_recv(client, telnet_recv, 0, NULL)) {
+		LOG_ERR("Unable to setup reception (family %u)",
+			net_context_get_family(client));
+		goto error;
+	}
+
+	LOG_DBG("Telnet client connected (family AF_INET%s)",
+		net_context_get_family(client) == AF_INET ? "" : "6");
+
+	sh_telnet->client_ctx = client;
+
+	return;
+error:
+	net_context_put(client);
+}
+
+static void telnet_setup_server(struct net_context **ctx, sa_family_t family,
+				struct sockaddr *addr, socklen_t addrlen)
+{
+	if (net_context_get(family, SOCK_STREAM, IPPROTO_TCP, ctx)) {
+		LOG_ERR("No context available");
+		goto error;
+	}
+
+	if (net_context_bind(*ctx, addr, addrlen)) {
+		LOG_ERR("Cannot bind on family AF_INET%s",
+			family == AF_INET ? "" : "6");
+		goto error;
+	}
+
+	if (net_context_listen(*ctx, 0)) {
+		LOG_ERR("Cannot listen on");
+		goto error;
+	}
+
+	if (net_context_accept(*ctx, telnet_accept, K_NO_WAIT, NULL)) {
+		LOG_ERR("Cannot accept");
+		goto error;
+	}
+
+	LOG_DBG("Telnet console enabled on AF_INET%s",
+		family == AF_INET ? "" : "6");
+
+	return;
+error:
+	LOG_ERR("Unable to start telnet on AF_INET%s",
+		family == AF_INET ? "" : "6");
+
+	if (*ctx) {
+		net_context_put(*ctx);
+		*ctx = NULL;
+	}
+}
+
+static int telnet_init(void)
+{
+#ifdef CONFIG_NET_IPV4
```

Comment:
Hmm, github lost my comment here, weird. Writing it again:
Better would use IS_ENABLED() here like this
```
if (IS_ENABLED(CONFIG_NET_IPV4)) {
	struct sockaddr_in any_addr4 = {
		.sin_family = AF_INET,
		.sin_port = htons(TELNET_PORT),
		.sin_addr = INADDR_ANY_INIT
	};
	static struct net_context *ctx4;

	telnet_setup_server(&ctx4, AF_INET,
			    (struct sockaddr *)&any_addr4,
			    sizeof(any_addr4));
}
```
and same for IPv6